ext/standard/array.c

```diff
@@ -4707,20 +4707,28 @@ PHP_FUNCTION(array_reduce)
 	}
 
 	fci.retval = &retval;
-	fci.param_count = 2;
+	fci.param_count = 3;
 	fci.no_separation = 0;
 
-	ZEND_HASH_FOREACH_VAL(htbl, operand) {
+	ZEND_HASH_FOREACH_BUCKET(htbl, operand) {
```

Comment:
You can use ZEND_HASH_FOREACH_KEY_VAL(ht, num, str, val) here, then if str == NULL it's a string key.